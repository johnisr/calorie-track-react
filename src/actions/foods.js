import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment';

export const addFood = food => ({
  type: 'ADD_FOOD',
  food: { id: uuid(), ...food },
});

export const startAddFood = (data = {}) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    const {
      name = '',
      amount = '0',
      unit = 'grams',
      carbohydrates = 0,
      protein = 0,
      fat = 0,
      calories = 0,
      timesUsed = 0,
      createdAt = moment(),
    } = data;
    const food = { name, amount, unit, carbohydrates, protein, fat, calories, timesUsed, createdAt };
    try {
      const res = await database.ref(`users/${uid}/foods`).push(food);
      dispatch(addFood({ id: res.key, ...food }));
    } catch (e) {

    }
  };
}

export const editFood = (id, updates) => ({
  type: 'EDIT_FOOD',
  id,
  updates,
});

export const startEditFood = (id, updates) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    try {
      await database.ref(`users/${uid}/foods/${id}`).update(updates);
      dispatch(editFood(id, updates));
    } catch (e) {
      
    }
  };
};

export const removeFood = ({ id }) => ({
  type: 'REMOVE_FOOD',
  id,
});

export const startRemoveFood = ({ id }) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    try {
      await database.ref(`users/${uid}/foods/${id}`).remove();
      dispatch(removeFood({ id }));
    } catch (e) {
      
    }
  };
};

export const setFoods = foods => ({
  type: 'SET_FOODS',
  foods,
});

export const startSetFoods = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    try {
      const snapshot = await database.ref(`users/${uid}/foods`).once('value');
      const foods = [];
      snapshot.forEach((food) => {
        foods.push({
          id: food.key,
          ...food.val(),
        });
      });
      dispatch(setFoods(foods));
    } catch (e) {

    }
  };
};