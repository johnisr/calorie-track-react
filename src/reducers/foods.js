const foodsDefaultState = [];

export default (state = foodsDefaultState, action) => {
  switch (action.type) {
    case 'ADD_FOOD':
      return [...state, action.food];
    case 'EDIT_FOOD':
      return state.map((food) => {
        if (food.id === action.id) {
          return {
            ...food,
            ...action.updates,
          };
        }
        return food;
      });
    case 'REMOVE_FOOD':
      return state.filter(({ id }) => id !== action.id);
    case 'SET_FOODS':
      return action.foods;
    default:
      return state;
  }
}