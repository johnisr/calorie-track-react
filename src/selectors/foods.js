export default (foods, { name, sortBy, maxFoodsShown, offset }) => {
  return foods
    .filter((food) => {
      const nameMatch = food.name.toLowerCase().includes(name.toLowerCase());
      return nameMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'usage') {
        return a.timesUsed < b.timesUsed ? 1 : -1;
      } else if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
    })
    .slice((offset * maxFoodsShown), (offset * maxFoodsShown) + maxFoodsShown);
}