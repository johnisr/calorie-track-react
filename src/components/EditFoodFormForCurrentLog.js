import React from 'react';
import { connect } from 'react-redux';
import FoodForm from './FoodForm';
import { editCurrentEditLog, editFoodFromCurrentLog, removeFoodFromCurrentLog } from '../actions/currentLog';

export class EditFoodFormForCurrentLog extends React.Component {
  handleSubmit = async (updates) => {
    await this.props.editFoodFromCurrentLog(this.props.index, updates);
    this.props.editCurrentEditLog({ editFoodIndex: undefined });
  }
  handleRemove = () => {
    this.props.editCurrentEditLog({ editFoodIndex: undefined });
    this.props.removeFoodFromCurrentLog(this.props.index);
  }
  handleExit = () => {
    this.props.editCurrentEditLog({ editFoodIndex: undefined });
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Food</h1>
          </div>
        </div>
          {
            (this.props.index !== undefined) ? (
              <div className="content-container">
                <FoodForm
                  food={this.props.food}
                  handleSubmit={this.handleSubmit}
                />
                <button onClick={this.handleRemove}>Remove</button>
                <button onClick={this.handleExit}>Exit</button>
                
              </div>
            ) : ( 
              <p>Select from List above to Edit Item</p>
            )
          }
      </div>
    )
  }
}


const mapStateToProps = (state, props) => ({
  index: state.currentLog.editFoodIndex,
  food: state.currentLog.editFoodIndex !== undefined ? state.currentLog.foods[state.currentLog.editFoodIndex] : undefined,
});
const mapDispatchToProps = dispatch => ({
  editCurrentEditLog: (updates) => dispatch(editCurrentEditLog(updates)),
  editFoodFromCurrentLog: (index, updates) => dispatch(editFoodFromCurrentLog(index, updates)),
  removeFoodFromCurrentLog: (index) => dispatch(removeFoodFromCurrentLog(index)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditFoodFormForCurrentLog);
