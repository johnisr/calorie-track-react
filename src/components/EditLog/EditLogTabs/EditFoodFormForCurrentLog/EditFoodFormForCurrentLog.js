import React from 'react';
import { connect } from 'react-redux';
import FoodForm from '../../../FoodForm/FoodForm';
import { editCurrentEditLog, editFoodFromCurrentLog, removeFoodFromCurrentLog } from '../../../../actions/currentLog';

export class EditFoodFormForCurrentLog extends React.Component {
  handleSubmit = async (updates) => {
    await this.props.editFoodFromCurrentLog(this.props.index, updates);
    this.props.editCurrentEditLog({ editFoodIndex: undefined });
  }
  handleRemove = () => {
    this.props.editCurrentEditLog({ editFoodIndex: undefined });
    this.props.removeFoodFromCurrentLog(this.props.index);
  }
  handleExit = () => {
    this.props.editCurrentEditLog({ editFoodIndex: undefined });
  }
  render() {
    return (
      <div className="pane">
        <h1 className="pane__title">Edit Food</h1>
        {
          (this.props.index !== undefined) ? (
            <div>
              <FoodForm
                food={this.props.food}
                handleSubmit={this.handleSubmit}
              />
              <div className="pane__buttons">
                <div>
                  <button className="btn btn--form" onClick={this.handleRemove}>Remove</button>
                </div>
                <div>
                  <button className="btn btn--form" onClick={this.handleExit}>Exit</button>
                </div>
              </div>
            </div>
          ) : ( 
            <p className="pane__warning">Select from List above to Edit Item</p>
          )
        }
      </div>
    )
  }
}


const mapStateToProps = (state, props) => ({
  index: state.currentLog.editFoodIndex,
  food: state.currentLog.editFoodIndex !== undefined ? state.currentLog.foods[state.currentLog.editFoodIndex] : undefined,
});
const mapDispatchToProps = dispatch => ({
  editCurrentEditLog: (updates) => dispatch(editCurrentEditLog(updates)),
  editFoodFromCurrentLog: (index, updates) => dispatch(editFoodFromCurrentLog(index, updates)),
  removeFoodFromCurrentLog: (index) => dispatch(removeFoodFromCurrentLog(index)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditFoodFormForCurrentLog);
