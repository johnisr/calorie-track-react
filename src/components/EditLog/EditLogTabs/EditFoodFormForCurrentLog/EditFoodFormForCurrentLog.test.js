import React from 'react';
import { shallow } from 'enzyme';
import { EditFoodFormForCurrentLog } from './EditFoodFormForCurrentLog';
import foods from '../../../../tests/fixtures/foods';

it('should render warning if no food to be edited', () => {
  wrapper = shallow(<EditFoodFormForCurrentLog
    index={undefined}
    food={undefined}
    editCurrentEditLog={jest.fn()}
    editFoodFromCurrentLog={jest.fn()}
    removeFoodFromCurrentLog={jest.fn()}
  />);
  expect(wrapper).toMatchSnapshot();
})

let wrapper, index, food, editCurrentEditLog, editFoodFromCurrentLog, removeFoodFromCurrentLog;
beforeEach(() => {
  index = 1;
  food = foods[1];
  editCurrentEditLog = jest.fn();
  editFoodFromCurrentLog = jest.fn();
  removeFoodFromCurrentLog = jest.fn();

  wrapper = shallow(<EditFoodFormForCurrentLog
    index={index}
    food={food}
    editCurrentEditLog={editCurrentEditLog}
    editFoodFromCurrentLog={editFoodFromCurrentLog}
    removeFoodFromCurrentLog={removeFoodFromCurrentLog}
  />);
});

it('should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should handleSubmit', async () => {
  const updates = {
    name: 'Chicken Thighs',
    amount: 8,
    unit: 'ounces',
    protein: 40,
    fat: 10,
    calories: 200,
  };
  await wrapper.find('FoodForm').prop('handleSubmit')(updates);
  expect(editFoodFromCurrentLog).toHaveBeenLastCalledWith(index, updates);
  expect(editCurrentEditLog).toHaveBeenLastCalledWith({ editFoodIndex: undefined });
});

it('should handleRemove', () => {
  wrapper.find('FoodForm').prop('onRemove')();
  expect(editCurrentEditLog).toHaveBeenLastCalledWith({ editFoodIndex: undefined });
  expect(removeFoodFromCurrentLog).toHaveBeenLastCalledWith(index);
});

it('should handleExt', () => {
  wrapper.find('FoodForm').prop('onExit')();
  expect(editCurrentEditLog).toHaveBeenLastCalledWith({ editFoodIndex: undefined });
});
