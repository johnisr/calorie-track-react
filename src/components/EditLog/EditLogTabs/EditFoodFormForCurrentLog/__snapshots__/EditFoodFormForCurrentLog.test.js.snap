// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditFoodFormForCurrentLog
    editCurrentEditLog={[Function]}
    editFoodFromCurrentLog={[Function]}
    food={
        Object {
            "amount": 100,
            "base": Object {
              "amount": 100,
              "calories": 165,
              "carbohydrates": 0,
              "fat": 3.6,
              "protein": 31,
            },
            "calories": 165,
            "carbohydrates": 0,
            "createdAt": 1000,
            "fat": 3.6,
            "id": "2",
            "name": "Chicken Breast",
            "protein": 31,
            "timesUsed": 1,
            "unit": "grams",
          }
    }
    index={1}
    removeFoodFromCurrentLog={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "food": Object {
        "amount": 100,
        "base": Object {
          "amount": 100,
          "calories": 165,
          "carbohydrates": 0,
          "fat": 3.6,
          "protein": 31,
        },
        "calories": 165,
        "carbohydrates": 0,
        "createdAt": 1000,
        "fat": 3.6,
        "id": "2",
        "name": "Chicken Breast",
        "protein": 31,
        "timesUsed": 1,
        "unit": "grams",
      },
      "handleSubmit": [Function],
      "onExit": [Function],
      "onRemove": [Function],
      "title": "Edit Food",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "food": Object {
          "amount": 100,
          "base": Object {
            "amount": 100,
            "calories": 165,
            "carbohydrates": 0,
            "fat": 3.6,
            "protein": 31,
          },
          "calories": 165,
          "carbohydrates": 0,
          "createdAt": 1000,
          "fat": 3.6,
          "id": "2",
          "name": "Chicken Breast",
          "protein": 31,
          "timesUsed": 1,
          "unit": "grams",
        },
        "handleSubmit": [Function],
        "onExit": [Function],
        "onRemove": [Function],
        "title": "Edit Food",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render warning if no food to be edited 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditFoodFormForCurrentLog
    editCurrentEditLog={[Function]}
    editFoodFromCurrentLog={[Function]}
    food={undefined}
    index={undefined}
    removeFoodFromCurrentLog={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Select from List above to Edit Item",
      "className": "center-text margin-bottom-small",
    },
    "ref": null,
    "rendered": "Select from List above to Edit Item",
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Select from List above to Edit Item",
        "className": "center-text margin-bottom-small",
      },
      "ref": null,
      "rendered": "Select from List above to Edit Item",
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
